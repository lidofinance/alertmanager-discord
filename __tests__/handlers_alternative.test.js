const axios = require("axios");
jest.mock("axios");

const { handleHook } = require("../handlers_alternative");

test("hook works", async () => {
  const getAlert = (i) => {
    return {
      status: "resolved",
      labels: { alertname: "activate" },
      annotations: {
        description: "cinema",
        summary: "donor" + String(i).padStart(2, "0"),
      },
    };
  };
  // mock
  const ctx = {
    routes: { test: "/dev/null" },
    params: { slug: "test" },
    request: {
      body: {
        alerts: Array.from({ length: 11 }, (_, i) => getAlert(i)),
      },
    },
  };

  axios.post.mockResolvedValue(null);

  await handleHook(ctx);

  expect(ctx.status).toBe(200);
  expect(axios.post.mock.calls.length).toBe(2);
  expect(axios.post.mock.calls).toMatchSnapshot();
});

test("balval features is correct", async () => {
  const getAlert = (i) => {
    return {
      status: i % 2 === 0 ? "resolved" : "fired", // resolved should be first
      labels: { alertname: "activate" },
      annotations: {
        description: "common description",
        resolved_description: "common resolved description",
        summary: "common summary with count",
        resolved_summary: "common resolved summary with count",
        field_name: "name" + String(i).padStart(2, "0"),
        field_value: "value" + String(i).padStart(2, "0"),
        url: `https://ya-${i % 2 === 0 ? "resolved" : "fired"}.ru`, // only last fired should have the same url
      },
    };
  };
  // mock
  const ctx = {
    routes: { test: "/dev/null" },
    params: { slug: "test" },
    request: {
      body: {
        alerts: Array.from({ length: 26 }, (_, i) => getAlert(i)),
      },
    },
  };

  axios.post.mockResolvedValue(null);

  await handleHook(ctx);

  expect(ctx.status).toBe(200);
  expect(axios.post.mock.calls.length).toBe(1);
  expect(axios.post.mock.calls).toMatchSnapshot();
});

test("mention works", async () => {
  const getAlert = (i) => {
    return {
      status: "resolved",
      labels: {
        alertname: "activate",
        mentions: i % 2 === 0 ? `mention-${i}, another-mention-${i}` : "",
      },
      annotations: {
        description: "cinema",
        summary: "donor" + String(i).padStart(2, "0"),
      },
    };
  };
  // mock
  const ctx = {
    routes: { test: "/dev/null" },
    params: { slug: "test" },
    request: {
      body: {
        alerts: Array.from({ length: 11 }, (_, i) => getAlert(i)),
      },
    },
  };

  axios.post.mockResolvedValue(null);

  await handleHook(ctx);

  expect(ctx.status).toBe(200);
  expect(axios.post.mock.calls.length).toBe(2);
  expect(axios.post.mock.calls).toMatchSnapshot();
});
