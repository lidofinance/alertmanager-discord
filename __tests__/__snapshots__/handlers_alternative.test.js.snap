// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`balval features are correct 1`] = `
Array [
  Array [
    "/dev/null",
    Object {
      "embeds": Array [
        Object {
          "color": 51283,
          "description": "common resolved description",
          "fields": Array [
            Object {
              "inline": true,
              "name": "name00",
              "value": "value00",
            },
            Object {
              "inline": true,
              "name": "name02",
              "value": "value02",
            },
            Object {
              "inline": true,
              "name": "name04",
              "value": "value04",
            },
            Object {
              "inline": true,
              "name": "name06",
              "value": "value06",
            },
            Object {
              "inline": true,
              "name": "name08",
              "value": "value08",
            },
            Object {
              "inline": true,
              "name": "name10",
              "value": "value10",
            },
            Object {
              "inline": true,
              "name": "name12",
              "value": "value12",
            },
            Object {
              "inline": true,
              "name": "name14",
              "value": "value14",
            },
            Object {
              "inline": true,
              "name": "name16",
              "value": "value16",
            },
            Object {
              "inline": true,
              "name": "name18",
              "value": "value18",
            },
            Object {
              "inline": true,
              "name": "name20",
              "value": "value20",
            },
            Object {
              "inline": true,
              "name": "name22",
              "value": "value22",
            },
            Object {
              "inline": true,
              "name": "name24",
              "value": "value24",
            },
          ],
          "footer": Object {
            "icon_url": "",
            "text": "",
          },
          "title": "13 common resolved summary with count",
          "url": "",
        },
        Object {
          "color": 13959168,
          "description": "common description",
          "fields": Array [
            Object {
              "inline": true,
              "name": "name01",
              "value": "value01",
            },
            Object {
              "inline": true,
              "name": "name03",
              "value": "value03",
            },
            Object {
              "inline": true,
              "name": "name05",
              "value": "value05",
            },
            Object {
              "inline": true,
              "name": "name07",
              "value": "value07",
            },
            Object {
              "inline": true,
              "name": "name09",
              "value": "value09",
            },
            Object {
              "inline": true,
              "name": "name11",
              "value": "value11",
            },
            Object {
              "inline": true,
              "name": "name13",
              "value": "value13",
            },
            Object {
              "inline": true,
              "name": "name15",
              "value": "value15",
            },
            Object {
              "inline": true,
              "name": "name17",
              "value": "value17",
            },
            Object {
              "inline": true,
              "name": "name19",
              "value": "value19",
            },
            Object {
              "inline": true,
              "name": "name21",
              "value": "value21",
            },
            Object {
              "inline": true,
              "name": "name23",
              "value": "value23",
            },
            Object {
              "inline": true,
              "name": "name25",
              "value": "value25",
            },
          ],
          "footer": Object {
            "icon_url": "",
            "text": "",
          },
          "title": "13 common summary with count",
          "url": "https://ya-fired.ru",
        },
      ],
    },
  ],
]
`;

exports[`hook works 1`] = `
Array [
  Array [
    "/dev/null",
    Object {
      "embeds": Array [
        Object {
          "color": 51283,
          "description": "cinema",
          "footer": Object {
            "icon_url": "",
            "text": "",
          },
          "title": "donor10",
          "url": "",
        },
        Object {
          "color": 51283,
          "description": "cinema",
          "footer": Object {
            "icon_url": "",
            "text": "",
          },
          "title": "donor09",
          "url": "",
        },
        Object {
          "color": 51283,
          "description": "cinema",
          "footer": Object {
            "icon_url": "",
            "text": "",
          },
          "title": "donor08",
          "url": "",
        },
        Object {
          "color": 51283,
          "description": "cinema",
          "footer": Object {
            "icon_url": "",
            "text": "",
          },
          "title": "donor07",
          "url": "",
        },
        Object {
          "color": 51283,
          "description": "cinema",
          "footer": Object {
            "icon_url": "",
            "text": "",
          },
          "title": "donor06",
          "url": "",
        },
        Object {
          "color": 51283,
          "description": "cinema",
          "footer": Object {
            "icon_url": "",
            "text": "",
          },
          "title": "donor05",
          "url": "",
        },
        Object {
          "color": 51283,
          "description": "cinema",
          "footer": Object {
            "icon_url": "",
            "text": "",
          },
          "title": "donor04",
          "url": "",
        },
        Object {
          "color": 51283,
          "description": "cinema",
          "footer": Object {
            "icon_url": "",
            "text": "",
          },
          "title": "donor03",
          "url": "",
        },
        Object {
          "color": 51283,
          "description": "cinema",
          "footer": Object {
            "icon_url": "",
            "text": "",
          },
          "title": "donor02",
          "url": "",
        },
        Object {
          "color": 51283,
          "description": "cinema",
          "footer": Object {
            "icon_url": "",
            "text": "",
          },
          "title": "donor01",
          "url": "",
        },
      ],
    },
  ],
  Array [
    "/dev/null",
    Object {
      "embeds": Array [
        Object {
          "color": 51283,
          "description": "cinema",
          "footer": Object {
            "icon_url": "",
            "text": "",
          },
          "title": "donor00",
          "url": "",
        },
      ],
    },
  ],
]
`;

exports[`mention works 1`] = `
Array [
  Array [
    "/dev/null",
    Object {
      "allowed_mentions": Object {
        "parse": Array [
          "users",
          "roles",
        ],
      },
      "content": "<@mention-10> <@another-mention-10> <@mention-8> <@another-mention-8> <@mention-6> <@another-mention-6> <@mention-4> <@another-mention-4> <@mention-2> <@another-mention-2>",
      "embeds": Array [
        Object {
          "color": 51283,
          "description": "cinema",
          "footer": Object {
            "icon_url": "",
            "text": "",
          },
          "title": "donor10",
          "url": "",
        },
        Object {
          "color": 51283,
          "description": "cinema",
          "footer": Object {
            "icon_url": "",
            "text": "",
          },
          "title": "donor09",
          "url": "",
        },
        Object {
          "color": 51283,
          "description": "cinema",
          "footer": Object {
            "icon_url": "",
            "text": "",
          },
          "title": "donor08",
          "url": "",
        },
        Object {
          "color": 51283,
          "description": "cinema",
          "footer": Object {
            "icon_url": "",
            "text": "",
          },
          "title": "donor07",
          "url": "",
        },
        Object {
          "color": 51283,
          "description": "cinema",
          "footer": Object {
            "icon_url": "",
            "text": "",
          },
          "title": "donor06",
          "url": "",
        },
        Object {
          "color": 51283,
          "description": "cinema",
          "footer": Object {
            "icon_url": "",
            "text": "",
          },
          "title": "donor05",
          "url": "",
        },
        Object {
          "color": 51283,
          "description": "cinema",
          "footer": Object {
            "icon_url": "",
            "text": "",
          },
          "title": "donor04",
          "url": "",
        },
        Object {
          "color": 51283,
          "description": "cinema",
          "footer": Object {
            "icon_url": "",
            "text": "",
          },
          "title": "donor03",
          "url": "",
        },
        Object {
          "color": 51283,
          "description": "cinema",
          "footer": Object {
            "icon_url": "",
            "text": "",
          },
          "title": "donor02",
          "url": "",
        },
        Object {
          "color": 51283,
          "description": "cinema",
          "footer": Object {
            "icon_url": "",
            "text": "",
          },
          "title": "donor01",
          "url": "",
        },
      ],
    },
  ],
  Array [
    "/dev/null",
    Object {
      "allowed_mentions": Object {
        "parse": Array [
          "users",
          "roles",
        ],
      },
      "content": "<@mention-0> <@another-mention-0>",
      "embeds": Array [
        Object {
          "color": 51283,
          "description": "cinema",
          "footer": Object {
            "icon_url": "",
            "text": "",
          },
          "title": "donor00",
          "url": "",
        },
      ],
    },
  ],
]
`;
